# 知識処理特論 最終課題

data: intelliProcess_last.txt (j101-d_11_1499.txt)

In [13]:
import numpy as np
from collections import Counter

In [2]:
f = open("../data/intelliProcess_last.txt")

text = f.read()
f.close()
text

'エレキベースのスラップ奏法を対象としたスペクトル特徴と演奏制約を利用した音高・奏法推定あらまし本論文では，エレキベースの奏法の違いによる音高の推定精度の低下を改善するために，特殊な奏法の一つであるスラップ奏法を対象にスペクトル特徴と演奏制約を導入し，ネットワークの最短経路を求めることで音高と奏法を推定する手法を提案する．キーワードエレキベース，調波打楽器音分離，スラップ奏法，非負値行列因子分解１．まえがき近年，計算機を用いて音楽信号から自動で楽譜を作成する自動採譜に関する研究が盛んに行われている［１］．自動採譜システムの実現のためには，楽曲の調やコードと関連が強いベースラインの推定が重要となる［２］．エレキベースの音高を知ることができれば，楽曲の調だけでなくコード進行を知ることができ，他の楽器音に対する採譜の手がかりともなる．エレキベースの奏法には，ピック弾き，指弾き，及びスラップ奏法と呼ばれる代表的な奏法がある．スラップ奏法はピック弾きや指弾きなどのスタンダードな奏法と比べて周波数特性が時間とともに大きく変化する．そのため，スタンダードな奏法とスラップ奏法とが混在している場合に，その周波数特性の異なりが音高推定を誤る原因の一つとなる．そこで，奏法の種類を推定して各音の高さの推定精度を向上させるために，音楽信号に対する自動採譜の効果的な手法の一つである非負値行列因子分解（ｎｏｎ−ｎｅｇａｔｉｖｅｍａｔｒｉｘｆａｃｔｏｒｉｚａｔｉｏｎ，ＮＭＦ）［３］に着目する．ＮＭＦによって，音楽信号のスペクトログラムを各音の周波数特性を表す基底スペクトルと各基底に対応する音の強さを表すアクティベーションとの積に近似的に分解できる．文献［４］ではＮＭＦの最適化にエレキベースの演奏の特徴を導入し，音高と発音時刻の推定精度を向上させる手法を提案しているが，演奏音にスラップ奏法を含む場合は想定されていない．スラップ奏法を含む楽曲に対して基底スペクトルの時不変を仮定しているＮＭＦを適用した場合，音高や発音時刻の推定精度が著しく低下してしまうという問題点があった．この問題を解決するために，本論文ではスラップ奏法のスペクトログラムに打楽器的な特徴があることに着目し，４弦のエレキベースのソロ演奏の音響信号に対して音高と奏法を推定する手法を提案する．２．提案手法エレキベースは弦とフレットの組み合わ

## 問１

述べ文字数，および異なり文字数を答えなさい

In [11]:
print("述べ文字数：" + str(len(text)))

diff_char = text[0]
for i in text:
    flag = 0
    for j in diff_char:
        if i == j:
            flag = 1
    if flag == 1:
        continue
    else:
        diff_char += i

print("異なり文字数：" + str(len(diff_char)))

述べ文字数：5960
異なり文字数：450


## 問２

エントロピーを求めることを考える．

(a) 全ての文字種が等確率で出現するとした時の（文字単位の）エントロピー H_0 を答えなさい．

(b) 文字の出現頻度（確率）を考慮した場合のエントロピー H_1 を答えなさい．

(c) 上記結果を比較し，その考察を記述しなさい．

### (a)

In [12]:
H0 = -np.log2(1/len(diff_char))
H0

8.813781191217037

$$
    H_0 = 8.813781191217037
$$

### (b)

In [19]:
x_list = []
x_list = Counter(text)
H1 = 0

for i in x_list:
    p = x_list[i]/len(text)
    H1 += -p*np.log2(p)

H1

7.459371410966394

$$
    H_1 = 7.459371410966394
$$

### (c)

(a)および(b)より，出現頻度を考慮したエントロピーH_1の方が値が小さい．エントロピーが小さいほど不確実性も低いため，H_1の方が優れた結果と言える．

## 問3

(a) 最も単純な方法として，コーパスに出現する全文字種に対してtri-gram統計を求めるとした場合に，必要となる具体的なバイト数を答えなさい．

(b) Nagaoらの手法で必要となる具体的なバイト数を答えなさい．

(c) 上記結果を比較し，その考察を記述しなさい．